### ISSUE TITLE
Optimize tensor operations with einsum for better performance

### ISSUE THREAD

**Issue by LiamConnell** [LiamConnell/FootyAI#6]
# Einsum Optimization Opportunities

After analyzing `src/v2/torch_soccer_env.py` and `src/v2/train.py`, I've identified several tensor operations that could benefit from `torch.einsum` for improved performance and memory efficiency.

### 1. Distance Calculations (torch_soccer_env.py:82-84, 124-126)
**Current:**
```python
team1_to_ball = torch.norm(self.team1_positions - self.ball_position.unsqueeze(1), dim=2)
team2_to_ball = torch.norm(self.team2_positions - self.ball_position.unsqueeze(1), dim=2)
```

**Potential einsum optimization:**
```python
# Calculate squared distances using einsum, then sqrt
diff = self.team1_positions - self.ball_position.unsqueeze(1)
team1_to_ball = torch.sqrt(torch.einsum('bni,bni->bn', diff, diff))
```
**Benefit:** Eliminates intermediate tensor creation and leverages optimized einsum kernel.

### 2. Masked Kick Summations (torch_soccer_env.py:134, 138)
**Current:**
```python
team1_total_kicks = torch.sum(team1_kick_mask * team1_kick_vel, dim=1)
team2_total_kicks = torch.sum(team2_kick_mask * team2_kick_vel, dim=1)
```

**Potential einsum optimization:**
```python
team1_total_kicks = torch.einsum('bni,bni->bi', team1_kick_mask, team1_kick_vel)
team2_total_kicks = torch.einsum('bni,bni->bi', team2_kick_mask, team2_kick_vel)
```
**Benefit:** More efficient broadcasting and reduction in a single operation.

### 3. Policy Loss Calculations (train.py:81-82)
**Current:**
```python
loss_a = - (log_probs_a.sum(dim=-1) * discounted_rewards_a).sum(dim=0).mean()
loss_b = - (log_probs_b.sum(dim=-1) * discounted_rewards_b).sum(dim=0).mean()
```

**Potential einsum optimization:**
```python
loss_a = -torch.einsum('tbi,tb->', log_probs_a, discounted_rewards_a) / batch_size
loss_b = -torch.einsum('tbi,tb->', log_probs_b, discounted_rewards_b) / batch_size
```
**Benefit:** Combines summation and element-wise multiplication in optimized kernel.

### 4. Kicker Counting (torch_soccer_env.py:135, 139)
**Current:**
```python
team1_kickers_count = torch.sum(team1_kick_mask, dim=(1, 2))
team2_kickers_count = torch.sum(team2_kick_mask, dim=(1, 2))
```

**Potential einsum optimization:**
```python
team1_kickers_count = torch.einsum('bni->b', team1_kick_mask)
team2_kickers_count = torch.einsum('bni->b', team2_kick_mask)
```
**Benefit:** More explicit about the reduction pattern, potentially better optimized.

### 5. Average Distance Calculations (torch_soccer_env.py:188-189)
**Current:**
```python
avg_dist_team1 = torch.mean(team1_to_ball, dim=1)
avg_dist_team2 = torch.mean(team2_to_ball, dim=1)
```

**Potential einsum optimization:**
```python
avg_dist_team1 = torch.einsum('bn->b', team1_to_ball) / N_PLAYERS
avg_dist_team2 = torch.einsum('bn->b', team2_to_ball) / N_PLAYERS
```
**Benefit:** More control over the reduction operation.

### 6. Discounted Rewards Calculation (train.py:46-50)
**Current:**
```python
discounted_temp = rewards_tensor * discounts
discounted_cumsum = torch.flip(torch.cumsum(torch.flip(discounted_temp, dims=[0]), dim=0), dims=[0])
discounted_rewards = discounted_cumsum / discounts
```

**Potential einsum optimization:** This is complex but could potentially use a custom einsum pattern for the discounting operation.

## Implementation Notes

- **Performance Testing Required:** Each optimization should be benchmarked to ensure actual performance gains
- **Memory Efficiency:** Einsum can reduce intermediate tensor allocations
- **Batch Size Sensitivity:** Benefits may vary with different batch sizes
- **GPU vs CPU:** Einsum optimizations often show greater benefits on GPU

## Priority Order
1. Distance calculations (most compute-intensive)
2. Masked kick summations (frequent operations)
3. Policy loss calculations (training bottleneck)
4. Other reductions and aggregations

Would benefit from profiling current performance before/after implementing these optimizations.
